src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala

```diff
@@ -0,0 +1,206 @@
+
+        val operationTimestamp = System.currentTimeMillis()
+        allFiles.map(_.removeWithTimestamp(operationTimestamp))
+      case Some(cond) =>
+        val (metadataPredicates, otherPredicates) =
+          DeltaTableUtils.splitMetadataAndDataPredicates(
+            cond, txn.metadata.partitionColumns, sparkSession)
+
+        if (otherPredicates.isEmpty) {
+          // Case 2: The condition can be evaluated using metadata only.
+          //         Delete a set of files without the need of scanning any data files.
+          val operationTimestamp = System.currentTimeMillis()
+          val candidateFiles = txn.filterFiles(metadataPredicates)
+
+          scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+          numTouchedFiles = candidateFiles.size
+
+          candidateFiles.map(_.removeWithTimestamp(operationTimestamp))
+        } else {
+          // Case 3: Delete the rows based on the condition.
+          val candidateFiles = txn.filterFiles(metadataPredicates ++ otherPredicates)
+
+          numTouchedFiles = candidateFiles.size
+          val nameToAddFileMap = generateCandidateFileMap(deltaLog.dataPath, candidateFiles)
+
+          val fileIndex = new TahoeBatchFileIndex(
+            sparkSession, "delete", candidateFiles, deltaLog, tahoeFileIndex.path, txn.snapshot)
+          // Keep everything from the resolved target except a new TahoeFileIndex
+          // that only involves the affected files instead of all files.
+          val newTarget = DeltaTableUtils.replaceFileIndex(target, fileIndex)
+          val data = Dataset.ofRows(sparkSession, newTarget)
+          val filesToRewrite =
+            withStatusCode("DELTA", s"Finding files to rewrite for DELETE operation") {
+              if (numTouchedFiles == 0) {
+                Array.empty[String]
+              } else {
+                data.filter(new Column(cond)).select(new Column(InputFileName())).distinct()
+                  .as[String].collect()
+              }
+            }
+
+          scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
+          if (filesToRewrite.isEmpty) {
+            // Case 3.1: no row matches and no delete will be triggered
+            Nil
+          } else {
+            // Case 3.2: some files need an update to remove the deleted files
+            // Do the second pass and just read the affected files
+            val baseRelation = buildBaseRelation(
+              sparkSession, txn, "delete", tahoeFileIndex.path, filesToRewrite, nameToAddFileMap)
+            // Keep everything from the resolved target except a new TahoeFileIndex
+            // that only involves the affected files instead of all files.
+            val newTarget = DeltaTableUtils.replaceFileIndex(target, baseRelation.location)
+
+            val targetDF = Dataset.ofRows(sparkSession, newTarget)
+            val filterCond = Not(EqualNullSafe(cond, Literal(true, BooleanType)))
```

Comment:
The  analyzer will check if the filter `hasNullAwarePredicateWithinNot`，like `Not( xxx in select yyy)`, will this pass the check, or, if not need we modify the check?

Reply:
Good observation. I dont think we can support subqueries in DELETE as of now since Apache Spark has limited support for subqueries. @ZhitongDB please add additional checks and tests to ensure that Delete throws a good error message regarding subqueries not being supported. These check can be done in DeltaCommand.apply before generate DeleteCommand from Delete.

Reply:
@tdas Do you mean we only support delete records according to specific conditions? But IMHO delete via subquery is a common use case. If we don't implement it today, we need implement it some days later.